ably/realtime_channel.go

```diff
@@ -2,130 +2,342 @@ package ably
+	return c.PublishAll([]*proto.Message{{Name: name, Data: data}}, listen)
+}
+
+// PublishAll publishes all given messages on the channel at once.
+// PublishAll does not block.
+//
+// If listen channel is non-nil it will be used to notify whether the messages
+// were published succesfully with nil error value; when the operation failed,
+// the error that caused the failure will be sent on the listen channel.
+//
+// This implicitly attaches the channel if it's not already attached.
+func (c *RealtimeChannel) PublishAll(messages []*proto.Message, listen chan<- error) error {
+	if err := c.Attach(); err != nil {
+		return err
+	}
 	msg := &proto.ProtocolMessage{
-		Action:  proto.ActionMessage,
-		Channel: c.Name,
-		Messages: []*proto.Message{
-			{Name: name, Data: data},
-		},
+		Action:   proto.ActionMessage,
+		Channel:  c.Name,
+		Messages: messages,
 	}
-	return c.client.send(msg)
+	switch c.State() {
+	case StateChanInitialized, StateChanAttaching:
+		if c.client.opts.NoQueueing {
+			return errQueueing
+		}
+		c.queue.Enqueue(msg, listen)
+		return nil
+	case StateChanAttached:
+	default:
+		return &Error{Code: 90001}
+	}
+	return c.client.Connection.send(msg, listen)
+}
+
+// State gives current state of the channel.
+func (c *RealtimeChannel) State() int {
+	c.state.Lock()
+	state := c.state.current
+	c.state.Unlock()
+	return state
+}
+
+// Reason gives the last error that caused channel transition to failed state.
+func (c *RealtimeChannel) Reason() error {
+	c.state.Lock()
+	err := c.state.err
+	c.state.Unlock()
+	return err
 }
 
 func (c *RealtimeChannel) notify(msg *proto.ProtocolMessage) {
 	switch msg.Action {
 	case proto.ActionAttached:
-		c.setState(ChanStateAttached)
+		c.state.syncSet(StateChanAttached, nil)
+		c.queue.Flush()
 	case proto.ActionDetached:
-		c.setState(ChanStateDetached)
+		c.state.syncSet(StateChanDetached, nil)
 	case proto.ActionPresence:
-		// TODO
+		Log.Printf(LogError, "TODO: implement RealtimePresence: %v", msg)
 	case proto.ActionError:
-		c.setState(ChanStateFailed)
-		c.Err = msg.Error
-		// TODO c.Close()
+		c.state.syncSet(StateChanFailed, newErrorProto(msg.Error))
+		c.queue.Fail(newErrorProto(msg.Error))
+		// TODO recovery
 	case proto.ActionMessage:
-		c.listenMtx.RLock()
-		defer c.listenMtx.RUnlock()
-		for _, m := range msg.Messages {
-			if l, ok := c.listeners[""]; ok {
+		c.listenMtx.Lock()
+		for _, msg := range msg.Messages {
+			if l, ok := c.listen[""]; ok {
 				for ch := range l {
-					ch <- m
+					select {
+					case ch <- msg:
+					default:
+						Log.Printf(LogWarn, "dropped %q message due to slow receiver", msg.Name)
+					}
 				}
 			}
-			if l, ok := c.listeners[m.Name]; ok {
+			if l, ok := c.listen[msg.Name]; ok {
 				for ch := range l {
-					ch <- m
+					select {
+					case ch <- msg:
+					default:
+						Log.Printf(LogWarn, "dropped %q message due to slow receiver", msg.Name)
+					}
 				}
 			}
```

Comment:
So I involved @mattheworiordan on this too, and I think that the main problem we have here is that we don't really want to drop any messages at all.

Since Go doesn't provide channels with infinite buffers (AFAIK), it sounds like we will have to go with queues instead.

Another point is that the Queue example that @lmars provided makes the queue private but doesn't define an accessor to know how big the queue is. It would be nice for users to have access to this information too in case they ever run into a situation where they have slow receivers.